firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java

```diff
@@ -4578,22 +4578,315 @@ public void testGetWaitsForConnection() throws DatabaseException, InterruptedExc
+    FirebaseApp writeApp =
+        appForDatabaseUrl(IntegrationTestValues.getDatabaseUrl(), UUID.randomUUID().toString());
+    FirebaseDatabase readDb = FirebaseDatabase.getInstance(readApp);
+    FirebaseDatabase writeDb = FirebaseDatabase.getInstance(writeApp);
+    Semaphore semaphore = new Semaphore(0);
+    long val = 34;
+    String parentNodeKey = Objects.requireNonNull(writeDb.getReference().push().getKey());
+    DatabaseReference parentWriteNode = writeDb.getReference().child(parentNodeKey);
+    String childNodeKey = Objects.requireNonNull(parentWriteNode.push().getKey());
+    DatabaseReference childReadNode =
+        readDb.getReference().child(parentNodeKey).child(childNodeKey);
+    DatabaseReference childWriteNode = parentWriteNode.child(childNodeKey);
+    ValueEventListener listener =
+        new ValueEventListener() {
+          @Override
+          public void onDataChange(@NonNull DataSnapshot snapshot) {
+            semaphore.release();
+          }
+
+          @Override
+          public void onCancelled(@NonNull DatabaseError error) {
+            // no-op
+          }
+        };
+
+    try {
+      await(childWriteNode.setValue(val));
+      ReadFuture.untilEquals(childReadNode, val).timedGet();
+      childReadNode.addValueEventListener(listener);
+      DatabaseReference parentReadNode = readDb.getReference().child(parentNodeKey);
+      IntegrationTestHelpers.waitFor(semaphore);
+      DataSnapshot snapshot = await(parentReadNode.get());
+      assertEquals(snapshot.child(childNodeKey).getValue(), val);
+      await(parentWriteNode.setValue(new MapBuilder().put(childNodeKey, val).build()));
+      DataSnapshot parentSnapshot = await(childReadNode.get());
+      assertEquals(parentSnapshot.getValue(), val);
+      childReadNode.removeEventListener(listener);
+    } catch (ExecutionException | TimeoutException | TestFailure e) {
+      fail("get threw an exception: " + e);
+    }
+  }
+
+  @Test
+  public void testLimitToGetWithListenerOnlyFiresOnce()
+      throws DatabaseException, InterruptedException, ExecutionException {
+    FirebaseApp readApp =
+        appForDatabaseUrl(IntegrationTestValues.getDatabaseUrl(), UUID.randomUUID().toString());
+    FirebaseApp writeApp =
+        appForDatabaseUrl(IntegrationTestValues.getDatabaseUrl(), UUID.randomUUID().toString());
+    FirebaseDatabase readDb = FirebaseDatabase.getInstance(readApp);
+    FirebaseDatabase writeDb = FirebaseDatabase.getInstance(writeApp);
+    String refKey = UUID.randomUUID().toString();
+    DatabaseReference writeRef = writeDb.getReference().child(refKey);
+    DatabaseReference readRef = readDb.getReference().child(refKey);
+    Map<String, Object> expected =
+        new MapBuilder().put("child1", "test1").put("child2", "test2").build();
+    try {
+      await(writeRef.setValue(expected));
+      ReadFuture.untilEquals(readRef, expected).timedGet();
```

Comment:
Was this test supposed to install a listener with a timeout and validate that we don't get two events in some window?

Reply:
I believe so